File: JavascriptAlgorithm/Tree/lowestCommonAncestor-236.js
var lowestCommonAncestor = function (root, p, q) {
  if (!root) {
    return null;
  }
  if (root.val === p || root.val === q) {
    return root;
  }
  let left = lowestCommonAncestor(root.left, p, q);
  let right = lowestCommonAncestor(root.right, p, q);
  if (!left && !right) {
    return null;
  }
  if (left && right) {
    return root;
  }
  return left || right;
};

export { lowestCommonAncestor };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/ThirdRound/Chapter07/lowestCommonAncestor_236.java
package algo.tzashinorpu.ThirdRound.Chapter07;

import algo.tzashinorpu.ThirdRound.TreeNode;

public class lowestCommonAncestor_236 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return root;
        // 当前树内是否有节点
        if (root == p || root == q) return root;
        // 左子树寻找这两个节点
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        // 右子树寻找这两个节点
        TreeNode right = lowestCommonAncestor(root.right, p, q);
//        return left == null ? (right == null ? root : right) : left;
        return left != null && right != null ? root : (left != null ? left : right);
    }
}
